internal/worker/worker.go

```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
-		if err != nil {
-			return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching traces for block %d: %v", blockNumber, err)}
+	for i, elem := range batch {
+		blockResults[i].BlockNumber = blockNumbers[i]
+
+		if elem.Error != nil {
+			blockResults[i].Error = elem.Error
+		} else {
+			result := *elem.Result.(*map[string]interface{})
+			serializedBlock := SerializeBlock(w.rpc.ChainID, result)
+			blockResults[i].Block = serializedBlock
+			serializedTransactions := SerializeTransactions(result["transactions"].([]interface{}), serializedBlock.Timestamp)
+			blockResults[i].Transactions = serializedTransactions
 		}
 	}
 
-	return SerializeBlockResult(w.rpc, block, logs, traces)
+	return blockResults
 }
 
-func (w *Worker) fetchBlock(blockNumber *big.Int) (*types.Block, error) {
-	return w.rpc.EthClient.BlockByNumber(context.Background(), blockNumber)
-}
+func (w *Worker) fetchLogs(blockNumbers []*big.Int) []LogsFetchResult {
+	if w.rpc.BlocksPerRequest.Logs >= len(blockNumbers) {
+		return w.fetchLogsBatch(blockNumbers)
+	}
+	chunks := divideIntoChunks(blockNumbers, w.rpc.BlocksPerRequest.Logs)
+	log.Debug().Msgf("Fetching logs with %d chunks", len(chunks))
 
-func (w *Worker) fetchLogs(blockNumber *big.Int) ([]types.Log, error) {
-	return w.rpc.EthClient.FilterLogs(context.Background(), ethereum.FilterQuery{
-		FromBlock: blockNumber,
-		ToBlock:   blockNumber,
-	})
-}
+	var wg sync.WaitGroup
+	resultsCh := make(chan []LogsFetchResult, len(chunks))
+
+	for _, chunk := range chunks {
+		wg.Add(1)
+		go func(chunk []*big.Int) {
+			defer wg.Done()
+			resultsCh <- w.fetchLogsBatch(chunk)
+		}(chunk)
+	}
+	go func() {
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	results := make([]LogsFetchResult, 0, len(blockNumbers))
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
+	}
 
-func (w *Worker) fetchTraces(blockNumber *big.Int) ([]map[string]interface{}, error) {
-	var result []map[string]interface{}
-	err := w.rpc.RPCClient.Call(&result, "trace_block", fmt.Sprintf("0x%x", blockNumber))
-	return result, err
+	return results
 }
 
-func (w *Worker) queryRows() {
-	// TODO: Implement this
-	/*rows, err := w.storage.DBStorage.Query(context.Background(), "SELECT version()")
+func (w *Worker) fetchLogsBatch(blockNumbers []*big.Int) []LogsFetchResult {
+	batch := make([]rpc.BatchElem, len(blockNumbers))
+	logsResults := make([]LogsFetchResult, len(blockNumbers))
+
+	for i, blockNum := range blockNumbers {
+		logsResults[i] = LogsFetchResult{
+			BlockNumber: blockNum,
+		}
+		hexBlockNum := hexutil.EncodeBig(blockNum)
+		batch[i] = rpc.BatchElem{
+			Method: "eth_getLogs",
+			Args:   []interface{}{map[string]string{"fromBlock": hexBlockNum, "toBlock": hexBlockNum}},
+			Result: new([]map[string]interface{}),
+		}
+	}
+
+	err := w.rpc.RPCClient.BatchCallContext(context.Background(), batch)
 	if err != nil {
-		log.Error().Err(err).Msg("Error querying ClickHouse")
+		for i := range logsResults {
```

Comment:
same here, does this mean one failed block mean all batch should fail?

Reply:
Same response as to the other one, so resolving this